service/history/historyEngine.go

```diff
@@ -645,14 +647,26 @@ func (e *historyEngineImpl) startWorkflowHelper(
 
 	workflowID := request.GetWorkflowId()
 	domainID := domainEntry.GetInfo().ID
+
 	// grab the current context as a lock, nothing more
+	// use a smaller context timeout to get the lock
+	getLockContext := ctx
+	var cancel context.CancelFunc
+	if deadline, ok := ctx.Deadline(); ok {
+		now := e.shard.GetTimeSource().Now()
+		ctxTimeout := now.Sub(deadline) / 2
+		if ctxTimeout > 0 {
+			getLockContext, cancel = context.WithTimeout(context.Background(), ctxTimeout)
+			defer cancel()
+		}
+	}
 	_, currentRelease, err := e.executionCache.GetOrCreateCurrentWorkflowExecution(
-		ctx,
+		getLockContext,
 		domainID,
 		workflowID,
 	)
 	if err != nil {
-		return nil, err
+		return nil, ErrConcurrentStartRequest
```

Comment:
I think I lack the context. So we decided to ignore the situation where the timeout is actually due to some DB issue instead of too many requests?

Reply:
Synced offline, we should enforce context timeout in DB layer so that the lock will be released on the context timedout.

Reply:
I think I am confused as well here. It looks like to me we try to get the in memory lock first and only if we successfully do that do we go on to update the database. So why do we have to enforce this context timeout at the db level?

Reply:
Also the error could be non-nil here because the execution cache is full technically. I think we should check that the error is of type context timeout here.

Reply:
So the release lock will be released after this function returns. Currently, the DB operations do not take in the context timeout. So in the worst case, it will hold this lock for 10 seconds. Within these 10 seconds, if the customers call this API with the same workflow ID, those requests could fail as a service busy error which will impact the availability metrics. So in the long run, we need the DB operations to take in context timeout